### ISSUE TITLE
box: box.session.on_disconnect() trigger has lost session id after yield

### ISSUE THREAD

**Issue by Totktonada** [tarantool/tarantool#4627]
Tarantool version: 2.3.0-195-gd56d869ad.
OS version: Linux.

```lua
#!/usr/bin/env tarantool

local net_box = require('net.box')
local fiber = require('fiber')

box.cfg{listen = 3301}

local latch = fiber.channel(1)
local sid_before_yield
local sid_after_yield
box.session.on_disconnect(function()
    sid_before_yield = box.session.id()
    fiber.yield()
    sid_after_yield = box.session.id()
    latch:put(true)
end)

local connection = net_box.connect('localhost:3301')
local res = connection:ping()
assert(res == true)
connection:close()

latch:get()
assert(sid_before_yield > 0)
assert(sid_before_yield == sid_after_yield)

os.exit()
```

`sid_before_yield` is 2, `sid_after_yield` is 0 on the tested tarantool version.

Reported by @Mons.

**Comment by Gerold103** [comment]
This is use after free. IProto thread sends two messages to TX: disconnect a session, and destroy it. If disconnect yields, destroy arrives to TX and gets handled before disconnect is done. After that the disconnect fiber works with a freed session.

**Comment by kostja** [comment]
I do not think disconnect triggers were ever allowed to yield. I believe
this is documented. If not, it should be.

**Comment by Totktonada** [comment]
If it is not allowed, then it should be forbidden: give an error likewise a yield in a transaction (#2528).

However I think it should be fixed. There is suspicion that this issue affects tarantool/queue: if it so (I'll check) it means that even our own code are not free of triggers that may yield.

**Comment by Gerold103** [comment]
Forbid to yield looks like a crutch. On disconnect was allowed to yield always, and we already have users who rely on that (Mons). We can't just forbid it now. It needs to be fixed properly.

**Comment by Totktonada** [comment]
Fixed in 1.10.4-52-gcf99b51cf, 2.2.1-124-g581f3674e and 2.3.0-236-g6da9d3953.
